### PR TITLE
EZP-29474: Fix for the hasAccess method when a policy has no limitations

### PR DESCRIPTION
**Pull request by mikadamczyk**
| Question           | Answer
| ------------------ | ------------------
| **JIRA issue**     | [EZP-29474](https://jira.ez.no/browse/EZP-29474)
| **Bug/Improvement**| yes
| **New feature**    |no
| **Target version** | `7.2`
| **BC breaks**      |no
| **Tests pass**     | yes
| **Doc needed**     |no

When Role is assigned to a user or a user group with limitations, all policies have these limitations. Even if by design they should have no limitation. This PR fixed hasAccess method to grant access when user has policy without limitations.

**TODO**:
- [x] Implement feature / fix a bug.
- [x] Implement tests.
- [x] Fix new code according to Coding Standards (`$ composer fix-cs`).
- [x] Ask for Code Review.

### CODE REVIEW

File: eZ/Publish/API/Repository/Tests/PermissionResolverTest.php [hunk 1]
```diff
@@ -249,6 +249,84 @@ public function testHasAccessLimited()
...
+            'role_' . __FUNCTION__,
+            [
+                ['module' => $module, 'function' => $function],
+            ]
+        );
+        $roleService = $repository->getRoleService();
+        $userService = $repository->getUserService();
+        // create user in root user group to avoid overlapping of existing policies and limitations
+        $user = $this->createUser('user', 'John', 'Doe', $userService->loadUserGroup(4));
+        $roleLimitation = $limitation instanceof Limitation\RoleLimitation ? $limitation : null;
```
Comment (alongosz) [inline eZ/Publish/API/Repository/Tests/PermissionResolverTest.php]: AFAICS both provided Limitations are instances of `Limitation\RoleLimitation`, so this is not needed. 

I'm assuming you wanted to provide a possibility to extend this provider in the future, but this breaks Liskov substitution principle a bit.

Please make 1st argument of this method as `RoleLimitation $roleLimitation` and remove this line. 

If we would want to add another type of limitation, which behaves differently, then we'd have to do it using another argument.

Comment (ViniTou) [reply eZ/Publish/API/Repository/Tests/PermissionResolverTest.php]: I was asking @mikadamczyk about it too, and he pointed me onto \eZ\Publish\API\Repository\Tests\PermissionResolverTest::testCanUserWithLimitations
is this case somewhat different from that?

Comment (mikadamczyk) [reply eZ/Publish/API/Repository/Tests/PermissionResolverTest.php]: @alongosz fixed

Comment (alongosz) [reply eZ/Publish/API/Repository/Tests/PermissionResolverTest.php]: > I was asking @mikadamczyk about it too, and he pointed me onto \eZ\Publish\API\Repository\Tests\PermissionResolverTest::testCanUserWithLimitations
is this case somewhat different from that?

@ViniTou no, it's me being a noob :) I probably need to fix it as well... when "there is free time" :joy:

**Comment by alongosz** [comment]
@mikadamczyk please retarget and rebase it to `7.2`

**Comment by andrerom** [comment]
To add to review notice, this seems to not be a bug but rather a feature request: As a Developer I want hasAccess to not return role limitations as limitations array if they don't apply to the given module/function, but instead return true if it detects it can.

I don't fully know the context of where this is used in Admin UI, but given the nature of hasAccess(), it will by design always return array if there are a limitations, so a possible better story would be to identify how if we can more widely use canUser in Admin UI to avoid this "issue" completely.

**Comment by alongosz** [comment]
@andrerom to reproduce this issue you can simply add `section / view` policy for Editor, create Editor user, log in as Editor, navigate to Content -> Content structure to display home content (or any Content) and see `UnauthorizedException ("User does not have access to 'view' 'section'")`. Without this assignment there's no exception. This is definitely a bug :-)

It is caused by the fact that [SectionService::loadSection](https://github.com/ezsystems/ezpublish-kernel/blob/v7.2.1/eZ/Publish/Core/Repository/SectionService.php#L179-L181) does not check these limitation. It doesn't make sense for that policy anyway. There are other similar examples for policies without limitations.

So the real questions are:
- should we improve in AdminUI [checking if user can view section](https://github.com/ezsystems/ezplatform-admin-ui/blob/v1.2.1/src/lib/Tab/LocationView/DetailsTab.php#L140)?
- should `loadSection` and others similar perform more sophisticated check than just `hasAccess('section', 'view') !== true`.

From my POV inheritance of limitations from Role for policies which are designed not to have limitations doesn't make sense, but if you know a use case for which this will not behave as it should (either by giving too much permissions or restricting access too much), please let us know.

**Comment by andrerom** [comment]
From the Customer side this is a bug, this PR in it's current form is a design change. I'm not saying we can't backport it, a user bug is a user bug, see below for what I'm saying or traying to say.


> should loadSection and others similar perform more sophisticated check than just hasAccess('section', 'view') !== true. // but what context should be given there?

1. In theory we should be able to change that loadSection..() calls to use:

```php
        if (!$this->repository->canUser('section', 'view', $spiSection)) {
            throw new UnauthorizedException('section', 'view');
        }
```

This should work as role limitation should return ACCESS_ABSTAIN which should not block the access from passing.

`loadSections()` should use this to filter the array, it will by default only return all or none as there are no limitations, however as the system is extendable this would be the cleanest, most future proof way.

2. I _do_ agree we can tweak hasAccess()to omit role limitation check, **when** it does not apply to the given module/function

The _when_ part here here is the tricky part, which is not covered in this PR yet, we will need to come up with a way to define it. Could be a setting to denote which modules are bound by role limitations, content modules is at least one of the modules which should not omit role limitation check by default.

And we'll need to add a BC note about this so people with custom modules becomes aware they now need to configure this if they have own modules that acts as extensions to content module.

**Comment by alongosz** [comment]
Hmm... ok, indeed 
> ```php
>        if (!$this->repository->canUser('section', 'view', $spiSection)) {
>            throw new UnauthorizedException('section', 'view');
>        }
> ```

also makes the error disappear. // cc @mikadamczyk 
I didn't think I even can use SPI object here ;)

I'm unsure however if this solves the actual problem. @andrerom the thing is, that I still don't see clearly a use case when limitation inherited from Role for `section / view` makes sense. Actually for any other policy defined [in the same way as `section / view`](https://github.com/ezsystems/ezpublish-kernel/blob/v7.2.1/eZ/Publish/Core/settings/policies.yml#L42).

On the other hand, we could argue that maybe Section limitation imposed on Role could be inherited. However when I configured it and applied the fix mentioned above, I still was able to load section that is not defined by the Role Limitation. Therefore I don't think it is supposed to work like this. If we want specific limitation to be inherited, we should allow this limitation for the policy itself.

Another argument in favor of current solution: from the technical standpoint `RoleLimitation` is a subtype of `Limitation` (the type which applies for policies). By inheriting policy Limitations from RoleLimitations we're trying to apply the same logic to parent type, which applies to subtype and in consequence this is not very SOLID.

So I really need an example to understand why you're against it and what it will break. I'm not saying I'm right. Permissions are critical part of the system and we cannot afford to make mistakes there. But I need more information on assumed business logic.

Moreover if we applied the above fix, there's no way to fix it in AdminUI, because we don't have context there. AdminUI does pre-load check, otherwise we'd need to refactor that into exception-driven logic, which is usually a mistake.

**Comment by andrerom** [comment]
@alongosz Ok, I'll take abit more time to explain things here, now since you seem to overlook what I'm trying to say.

The current approach, is assuming Role Assignment Limitation and Policy Limitations are the same thing, they are not! They are two disconnected things, which just happens to use the same Limitation Types system:

permissions


This PR will cause the following functions to no longer take Role Assignment Limitation into account:
```
parameters:
    ezpublish.api.role.policy_map:
        content:
            reverserelatedlist: ~
            translations: ~
            urltranslator: ~
            pendinglist: ~
            restore: ~
            cleantrash: ~
```

This might or might not be a security issue for customers that rely on this behavior.

It will also perhaps break user/* functions that pass user (content) object as context as they will then take role limitation into account, we do this on `!$this->permissionResolver->canUser('user', 'password', $loadedUser)` today it seems. _(Open question if this usage is a bug or not though, but that is besides the point here)_

So see section 2 above, as I said there we are missing some feature to make the connection and the assumptions you want to make here.

**Comment by mikadamczyk** [comment]
Now PR affects only section/view limitations. Subtree and Section limitations for policy section/view is marked as "not included".

**Comment by gggeek** [comment]
Going off a bit on a tangent here, but I always thought that the role-assignment-limitations were a bit weird in eZP, unlike policy limitations.

The first question that I have is: who/what decides which limitations are available as role-assignment-limitations ? In eZ4 we have subtree and siteaccess, but more might make sense.

A more interesting question is: how do the policies themselves (and their application) cope with role-assignment-limitations? Eg. if a policy is user/login, it makes no sense at 1st sight that it should care about an added limitation for a subtree. But should content/edit take care of an added siteaccess limitation?

This question, besides the theoretical importance of keeping the system sound, has the practical implication that we have to decide if it is the cms 'kernel' that takes care of 'adding' the role-assignment-limitation to the existing policy, or if it is the developer who is writing the implementation of the policy that has to take care about the possibility of that happening.

Going forward, an option could be to remove r-a-l altogether => although nice from a logical pov, we would need to make every single policy accept subtree/section/siteaccess limitation (at the moment only content/edit is that flexible), and have a gui that allows for easy mass assign of limittaions, since this is, in my experience, the current reason d'etre of the r-a-l

File: eZ/Publish/Core/settings/policies.yml [hunk 2]
```diff
@@ -39,7 +39,7 @@ parameters:
         section:
             assign: { Class: true, Section: true, Owner: true, NewSection: true }
             edit: ~
-            view: ~
+            view: { Subtree: false, Section: false }
```
Comment (andrerom) [inline eZ/Publish/Core/settings/policies.yml]: It works, and does not negatively affect other policies, but:
1. this is config for policy limitations, not role limitations
2. the issue QA reported here is probably present in several parts of the system, basically anything using hasAccess() and not expecting limitations back even tough the system is extendable so someone might add it

**Comment by Nattfarinn** [comment]
@andrerom But is there any reason why content policies ie.:

```
            urltranslator: ~
            pendinglist: ~
            restore: ~
            cleantrash: ~
```

don't have own limitations like Subtree and such, if these limitations can be implicit inherited by RoleAssignments anyway?

Why don't we make inheritance explicit by limiting it only to limitations that exists in both, RoleAssignment and Policy? What I'm trying to say, if policy can't have own limitations (or particular limitation) it shouldn't be inherited from RoleAssignment at all. If you want to inherit limitation from RoleAssignment by Role, make sure the Role policy itself can inherit such limitation.

This:
```
            urltranslator: { Subtree: true, Section: true }
            pendinglist: { Subtree: true, Section: true }
            restore: { Subtree: true, Section: true }
            cleantrash: { Subtree: true, Section: true }
            something: { Section: true }
            somethingelse: ~
```
makes much more sense to me. `something` policy could be limited (and would inherit from RoleAssignment) by Section limitation, `somethingelse` could not have and would not inherit any limitations while rest could be limited by Subtree and Section limitation and would inherit both.

> They are two disconnected things, which just happens to use the same Limitations values and Type system.

Maybe they are code base disconnected, yet they both are responsible by exact same thing but on two different levels. It's hard to not think about them as related to each other if we're discussing something what is indistinguishable from hierarchical and implicit inheritance.

**Comment by andrerom** [comment]
@Nattfarinn That config is for policy limitations, role limitations are a separate (but similar) concept as mentioned above.

**Comment by Nattfarinn** [comment]
@andrerom But Role itself is made from policies with limitations. Even as different limitation layers I would say they are more hierarchical than just separate but similar concepts, and I don't mind having one to rely on another, considering their purpose is the same but on different levels.

**Comment by andrerom** [comment]
So not saying you are off here, but for me there are two things here:
- The issue reported by QA, this is caused by wrong usage of API. I'm trying to fix that here: https://github.com/ezsystems/ezpublish-kernel/pull/2413 
- This feature request to expose some kind of new or improved hasAccess like API where we 1. don't need to provide object and 2. can safely know when we can omit certain policies _(for instance getting them to tell permission system they abstain from "voting" and can be ignored, aka does not need to be returned by API, see comment in PR above)_.

**Pull request closed without merging**